File: demos/animation-demo/src/main/java/org/pushingpixels/radiance/demo/animation/swing/tracker/BallPanel.java
/*
 * Copyright (c) 2005-2022 Radiance Kirill Grouchnikov. All Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without 
 * modification, are permitted provided that the following conditions are met:
 * 
 *  o Redistributions of source code must retain the above copyright notice, 
 *    this list of conditions and the following disclaimer. 
 *     
 *  o Redistributions in binary form must reproduce the above copyright notice, 
 *    this list of conditions and the following disclaimer in the documentation 
 *    and/or other materials provided with the distribution. 
 *     
 *  o Neither the name of the copyright holder nor the names of
 *    its contributors may be used to endorse or promote products derived 
 *    from this software without specific prior written permission. 
 *     
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, 
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR 
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR 
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, 
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; 
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE 
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, 
 * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
 */
package org.pushingpixels.radiance.demo.animation.swing.tracker;

import javax.swing.*;
import java.awt.*;
import java.awt.geom.Ellipse2D;

public class BallPanel extends JComponent {
    private float ballY;

    public static final int RADIUS = 20;

    public BallPanel() {
        this.ballY = RADIUS;
    }

    public void setBallY(float ballY) {
        this.ballY = ballY;
    }

    public float getBallY() {
        return ballY;
    }

    @Override
    protected void paintComponent(Graphics g) {
        Graphics2D g2d = (Graphics2D) g.create();

        g2d.setColor(Color.gray);
        g2d.fillRect(0, 0, getWidth(), getHeight());

        g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);

        Shape shape = new Ellipse2D.Double(this.getWidth() / 2 - RADIUS, this.ballY - RADIUS,
                2 * RADIUS, 2 * RADIUS);
        g2d.setPaint(new RadialGradientPaint(this.getWidth() / 2 - RADIUS / 4,
                this.ballY - RADIUS / 5, RADIUS + RADIUS / 4, new float[] { 0.0f, 1.0f },
                new Color[] { Color.green, Color.green.darker().darker() }));
        g2d.fill(shape);
        g2d.setColor(new Color(0, 64, 0));
        g2d.draw(shape);

        g2d.dispose();

    }
}

File: demos/torch-demo/src/main/kotlin/org/pushingpixels/demo/torch/swing/tracker/BallPanel.kt
/*
 * Copyright (c) 2005-2021 Radiance Kirill Grouchnikov. All Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without 
 * modification, are permitted provided that the following conditions are met:
 * 
 *  o Redistributions of source code must retain the above copyright notice, 
 *    this list of conditions and the following disclaimer. 
 *     
 *  o Redistributions in binary form must reproduce the above copyright notice, 
 *    this list of conditions and the following disclaimer in the documentation 
 *    and/or other materials provided with the distribution. 
 *     
 *  o Neither the name of the copyright holder nor the names of
 *    its contributors may be used to endorse or promote products derived 
 *    from this software without specific prior written permission. 
 *     
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, 
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR 
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR 
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, 
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; 
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE 
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, 
 * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
 */
package org.pushingpixels.demo.torch.swing.tracker

import org.pushingpixels.meteor.awt.deriveByBrightness
import org.pushingpixels.meteor.awt.render
import java.awt.Color
import java.awt.Graphics
import java.awt.RadialGradientPaint
import java.awt.RenderingHints
import java.awt.geom.Ellipse2D
import javax.swing.JComponent

class BallPanel : JComponent() {
    var ballY = RADIUS

    override fun paintComponent(g: Graphics) {
        g.render {
            it.color = Color.gray
            it.fillRect(0, 0, width, height)

            it.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON)

            val shape = Ellipse2D.Double((this.width / 2 - RADIUS).toDouble(),
                    (this.ballY - RADIUS).toDouble(),
                    (2 * RADIUS).toDouble(), (2 * RADIUS).toDouble())
            it.paint = RadialGradientPaint(this.width / 2 - RADIUS / 4,
                    this.ballY - RADIUS / 5, RADIUS + RADIUS / 4,
                    floatArrayOf(0.0f, 1.0f),
                    arrayOf(Color.green, Color.green.deriveByBrightness(-0.7)))
            it.fill(shape)
            it.color = Color(0, 64, 0)
            it.draw(shape)
        }
    }

    companion object {
        val RADIUS = 20.0f
    }
}
